src/commands.rs

```diff
@@ -205,149 +198,139 @@ impl Commands {
-    if i > 0 {
-        let mut char_set = CharacterSet::from_char(' ');
-        char_set.insert('\n');
+            state = self.state_machine.add(state, char_set);
+            self.state_machine.add_next_state(state, state);
+        }
+        state
+    }
+
+    fn add_help_menu(&mut self, cmd: &'static str, mut state: usize) -> usize {
+        "?help".chars().for_each(|ch| {
+            state = self.state_machine.add(state, CharacterSet::from_char(ch));
+        });
+        state = self.add_space(state, 1);
+        cmd.chars().for_each(|ch| {
+            state = self.state_machine.add(state, CharacterSet::from_char(ch));
+        });
 
-        state = state_machine.add(state, char_set);
-        state_machine.add_next_state(state, state);
+        state
     }
-    state
-}
 
-fn add_help_menu<T>(
-    mut state_machine: &mut StateMachine<T>,
-    cmd: &'static str,
-    mut state: usize,
-) -> usize {
-    "?help".chars().for_each(|ch| {
-        state = state_machine.add(state, CharacterSet::from_char(ch));
-    });
-    state = add_space(&mut state_machine, state, 1);
-    cmd.chars().for_each(|ch| {
-        state = state_machine.add(state, CharacterSet::from_char(ch));
-    });
-
-    state
-}
+    fn add_dynamic_segment(&mut self, mut state: usize, s: &'static str) -> usize {
+        let name = &s[1..s.len() - 1];
 
-fn add_dynamic_segment<T>(
-    state_machine: &mut StateMachine<T>,
-    name: &'static str,
-    mut state: usize,
-) -> usize {
-    let mut char_set = CharacterSet::any();
-    char_set.remove(' ');
-    state = state_machine.add(state, char_set);
-    state_machine.add_next_state(state, state);
-    state_machine.start_parse(state, name);
-    state_machine.end_parse(state);
-
-    state
-}
+        let mut char_set = CharacterSet::any();
+        char_set.remove(' ');
+        state = self.state_machine.add(state, char_set);
+        self.state_machine.add_next_state(state, state);
+        self.state_machine.start_parse(state, name);
+        self.state_machine.end_parse(state);
 
-fn add_remaining_segment<T>(
-    state_machine: &mut StateMachine<T>,
-    name: &'static str,
-    mut state: usize,
-) -> usize {
-    let char_set = CharacterSet::any();
-    state = state_machine.add(state, char_set);
-    state_machine.add_next_state(state, state);
-    state_machine.start_parse(state, name);
-    state_machine.end_parse(state);
-
-    state
-}
+        state
+    }
 
-fn add_code_segment_multi_line<T>(
-    state_machine: &mut StateMachine<T>,
-    name: &'static str,
-    mut state: usize,
-) -> usize {
-    state = state_machine.add(state, CharacterSet::from_char('`'));
-    state = state_machine.add(state, CharacterSet::from_char('`'));
-    state = state_machine.add(state, CharacterSet::from_char('`'));
+    fn add_remaining_segment(&mut self, mut state: usize, s: &'static str) -> usize {
+        let name = &s[..s.len() - 3];
```

Comment:
Is this guaranteed to operate on character boundaries? Indexes are on bytes.

Reply:
Yeah its fine, the only characters that fit in those states are ascii.

Reply:
We have been doing this indexing all along, i just moved it